feat(react): react server support (#246)
spotify/confidence-sdk-js: packages/react/README.md

### Initializing the ConfidenceProvider

The Confidence React integration has a Provider that needs to be initialized. It accepts a Confidence instance and should wrap your component tree. Here's an example for client-side rendering:

```ts
import { Confidence } from '@spotify-confidence/sdk/';

> import { ConfidenceProvider } from '@spotify-confidence/sdk/react/client';

Author: ```suggestion
import { ConfidenceProvider } from '@spotify-confidence/sdk/react';
```

}
```

For server-side rendering setup, see the [Server-Side Rendering Support](#server-side-rendering-support) section below.

Anywhere in the sub-tree under the `ConfidenceProvider` you can now access the confidence instance with the `useConfidence()` hook to access context modification API's. For flag resolves we suggest using `useFlag()`.

### Managing context

The `ConfidenceProvider` API supports a `useWithContext()` hook to achieve the [standard context API](https://github.com/spotify/confidence-sdk-js/blob/main/packages/sdk/README.md#setting-the-context).

### Accessing flags

Flags are accessed with a set of hooks exported from `@spotify-confidence/react`:

```ts
import { useFlag, useEvaluateFlag } from '@spotify-confidence/react';

function MyComponent() {
  // Simple flag access - returns the flag value or default
  const color = useFlag('my-feature-flag.color', 'blue');

  // Detailed flag evaluation - returns evaluation details

>   const { value, reason, context } = useEvaluateFlag('my-feature-flag.size', 12);

Reviewer: ```suggestion
  const { value, reason } = useEvaluateFlag('my-feature-flag.size', 12);
```

}
```

For server-side rendering setup, see the [Server-Side Rendering Support](#server-side-rendering-support) section below.

> Anywhere in the sub-tree under the `ConfidenceProvider` you can now access the confidence instance with the `useConfidence()` hook. The hook actually returns an instance of `ConfidenceReact` which is a wrapper around the normal `Confidence` API with some slight adaptations to integrate better with React. You can read more about the differences in the following sections.

Author: > The hook actually returns an instance of `ConfidenceReact`

Not true anymore.

- `useFlag(flagName, defaultValue)` will return the flag value or default.
- `useEvaluateFlag(flagName, defaultValue)` will return more details about the flag evaluation, together with the value.

Both of the flag hooks integrate with the React Suspense API so that the suspense fallback will be shown until flag values are available. It is therefore important to wrap any component using the above hooks in a suspense boundary.

> The hooks are also reactive so that if the context changes, any components using the hooks will be re-rendered. As dependent components are re-rendered as soon as the context changes, one might expect that would again trigger the suspense boundary while the flag values are resolved.

Author: > one might expect that would again trigger the suspense boundary while the flag values are resolved.
That sentence without the followup is little strange 😄

Reviewer: I had AI rewrite the main part of the access flags. let me know what you think.

> 

Reviewer: @andreas-karlsson can you check [L69](https://github.com/spotify/confidence-sdk-js/pull/246/files#diff-3dae6d110b675ef311b837cf10653c25945e787593bfc47d7bdbc3836df15abbR69)?

If the hooks can't resolve the flag values withing the timeout specified on the Confidence instance, they will instead return the default value.

### Server-Side Rendering Support (experimental)

> The Confidence React SDK now supports server-side rendering (SSR) and React Server Components (RSC) for instance in Next.js. The SDK provides separate entry points for client and server components:

Author: ```suggestion
The Confidence React SDK now supports server-side rendering (SSR) and React Server Components (RSC) for instance in Next.js. The SDK provides a separate entry point for server components:
```

> - `@spotify-confidence/react/client` - For client components

Author: ```suggestion
- `@spotify-confidence/react` - For client components
```

- `@spotify-confidence/react/server` - For server components

When using the SDK in a server environment:

1. Create a Confidence instance for server using the React.cache as the scope in CacheOptions.

> 2. Provide an accessor for the Confidence instance using `withContext`.

Author: This is a bit cryptic?

3. Use direct flag evaluation in server components.

Here's an example of how to use Confidence in a Next.js application:

```ts
// app/confidence.ts (Server-side configuration)
import { Confidence } from '@spotify-confidence/sdk';
import React from 'react';

const confidence = Confidence.create({
  clientSecret: process.env.CONFIDENCE_CLIENT_SECRET!,
  environment: 'backend',
  timeout: 1000,
  logger: console,
  cache: {
    scope: React.cache, // Use React.cache for server-side caching
  },
});

// Confidence accessor for use in RSC.
export const getConfidence = (context: Context): Confidence => {
  return confidence.withContext(context);

> };

Author: We don't need this step. We could recommend to just export the root instance and then do withContext(), maybe that'd be easier to document?

Reviewer: fixed
